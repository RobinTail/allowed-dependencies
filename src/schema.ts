import type { JSONSchema } from "@typescript-eslint/utils";
import type { FromSchema } from "json-schema-to-ts";
import { fromPairs, xprod } from "ramda";

const value = {
  oneOf: [{ type: "boolean" }, { type: "string", enum: ["typeOnly"] }],
} as const satisfies JSONSchema.JSONSchema4;
export type Value = FromSchema<typeof value>;

const categories = {
  type: "object",
  properties: fromPairs(
    xprod(["production", "optionalPeers", "requiredPeers"] as const, [value]),
  ),
} as const satisfies JSONSchema.JSONSchema4;
export type Category = keyof typeof categories.properties;

const manifest = {
  type: "object",
  properties: fromPairs(
    xprod(
      ["dependencies", "peerDependencies", "peerDependenciesMeta"] as const,
      [{ type: "object" } as const],
    ),
  ),
} as const satisfies JSONSchema.JSONSchema4;

export const options = {
  type: "object",
  properties: {
    manifest: manifest,
    typeOnly: { type: "array", items: { type: "string" } },
    ignore: { type: "array", items: { type: "string", format: "regex" } },
    ...categories.properties,
  },
  additionalProperties: false,
  required: ["manifest"],
} as const satisfies JSONSchema.JSONSchema4;
export type Options = FromSchema<typeof options>;
