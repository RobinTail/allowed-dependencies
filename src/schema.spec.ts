import { describe, expect, it } from "vitest";
import { options } from "./schema";

describe("Schema", () => {
  describe("options", () => {
    it("has the required props", () => {
      expect(options).toMatchSnapshot();
    });
  });
});
