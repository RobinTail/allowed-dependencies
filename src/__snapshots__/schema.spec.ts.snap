// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Schema > options > has the required props 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "ignore": {
      "items": {
        "format": "regex",
        "type": "string",
      },
      "type": "array",
    },
    "optionalPeers": {
      "oneOf": [
        {
          "type": "boolean",
        },
        {
          "enum": [
            "typeOnly",
          ],
          "type": "string",
        },
      ],
    },
    "packageDir": {
      "type": "string",
    },
    "production": {
      "oneOf": [
        {
          "type": "boolean",
        },
        {
          "enum": [
            "typeOnly",
          ],
          "type": "string",
        },
      ],
    },
    "requiredPeers": {
      "oneOf": [
        {
          "type": "boolean",
        },
        {
          "enum": [
            "typeOnly",
          ],
          "type": "string",
        },
      ],
    },
    "typeOnly": {
      "items": {
        "type": "string",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;
